File: elixir/anagram.ex
defmodule Anagram do
  @doc """
  Returns all candidates that are anagrams of, but not equal to, 'base'.
  """
  @spec match(String.t(), [String.t()]) :: [String.t()]
  def match(base, candidates) do
    base = String.downcase(base)
    sorted_base = sort(base)

    candidates
    |> Enum.filter(fn c ->
      c = String.downcase(c)
      c != base && sort(c) == sorted_base
    end)
  end

  defp sort(s) do
    s |> String.to_charlist() |> Enum.sort() |> List.to_string()
  end
end

File: ocaml/anagram.ml
open Base

let sort s =
  let lst = String.to_list s in
  List.sort lst Char.compare |> String.of_list

let anagrams s lst =
  let s = String.lowercase s in
  let ss = sort s in
  List.filter
    ~f:(fun t ->
      let t = String.lowercase t in
      let tt = sort t in
      (not (String.equal s t)) && String.equal ss tt)
    lst
